part/node.go

```diff
@@ -258,47 +263,154 @@ func (n *header[T]) children() []*header[T] {
+			return children[i].prefix[0] >= key
+		})
+		if idx >= n.size() || children[idx].prefix[0] != key {
+			// No node found, return nil and the index into
+			// which it should go.
+			return nil, idx
+		}
+		return children[idx], idx
+	case nodeKind256:
 		return n.node256().children[key], int(key)
+	default:
+		panic(fmt.Sprintf("unknown node kind: %x", n.kind()))
 	}
-
-	children := n.children()
-	idx := sort.Search(len(children), func(i int) bool {
-		return children[i].prefix[0] >= key
-	})
-	if idx >= n.size() || children[idx].prefix[0] != key {
-		// No node found, return nil and the index into
-		// which it should go.
-		return nil, idx
-	}
-	return children[idx], idx
 }
 
 func (n *header[T]) find(key byte) *header[T] {
-	child, _ := n.findIndex(key)
-	return child
+	switch n.kind() {
+	case nodeKindLeaf:
+		return nil
+	case nodeKind4:
+		n4 := n.node4()
+		size := n4.size()
+		for i := 0; i < int(size); i++ {
+			if n4.keys[i] == key {
+				return n4.children[i]
+			} else if n4.keys[i] > key {
+				return nil
+			}
+		}
+		return nil
+	case nodeKind16:
+		n16 := n.node16()
+		size := n16.size()
+		for i := 0; i < int(size); i++ {
+			if n16.keys[i] == key {
+				return n16.children[i]
+			} else if n16.keys[i] > key {
+				return nil
+			}
+		}
+		return nil
+	case nodeKind48:
+		n48 := n.node48()
+		idx := n48.index[key]
+		if idx < 0 {
+			return nil
+		}
+		return n48.children[idx]
+	case nodeKind256:
+		return n.node256().children[key]
+	default:
+		panic("unknown node kind")
+	}
 }
 
 func (n *header[T]) insert(idx int, child *header[T]) {
-	n.setSize(n.size() + 1)
+	size := n.size()
+	newSize := size + 1
 	switch n.kind() {
+	case nodeKind4:
+		n4 := n.node4()
+		// Shift to make room
+		copy(n4.children[idx+1:newSize], n4.children[idx:newSize])
+		copy(n4.keys[idx+1:newSize], n4.keys[idx:newSize])
+		n4.children[idx] = child
+		n4.keys[idx] = child.prefix[0]
+	case nodeKind16:
+		n16 := n.node16()
+		// Shift to make room
+		copy(n16.children[idx+1:newSize], n16.children[idx:newSize])
+		copy(n16.keys[idx+1:newSize], n16.keys[idx:newSize])
+		n16.children[idx] = child
+		n16.keys[idx] = child.prefix[0]
+	case nodeKind48:
+		// Shift to make room
+		n48 := n.node48()
+		for i := size - 1; i >= idx; i-- {
+			c := n48.children[i]
+			n48.index[c.prefix[0]] = int8(i + 1)
+			n48.children[i+1] = c
+		}
+		n48.children[idx] = child
+		n48.index[child.prefix[0]] = int8(idx)
+
 	case nodeKind256:
 		n.node256().children[child.prefix[0]] = child
-	default:
-		children := n.children()
-		// Shift to make room
-		copy(children[idx+1:], children[idx:])
-		children[idx] = child
 	}
+	n.setSize(size + 1)
 }
 
 func (n *header[T]) remove(idx int) {
-	if n.kind() == nodeKind256 {
-		n.node256().children[idx] = nil
-	} else {
+	switch n.kind() {
+	case nodeKind4:
+		size := n.size()
+		n4 := n.node4()
+		copy(n4.keys[idx:size], n4.keys[idx+1:size])
+		copy(n4.children[idx:size], n4.children[idx+1:size])
+		n4.children[n4.size()-1] = nil
```

Comment:
nit: many `size()-1` uses - how about `newSize := n.size() - 1` like in `insert`?